docs/notebooks/advanced_data_transformation.pct.py

```diff
@@ -0,0 +1,231 @@
+query_point_transformer = MinMaxTransformer(tf.stack((search_space.lower, search_space.upper)))
+observation_transformer = StandardTransformer(initial_data.observations)
+normalized_data = Dataset(
+    query_point_transformer.transform(initial_data.query_points),
+    observation_transformer.transform(initial_data.observations)
+)
+
+@transform_data(
+    query_point_transformer, observation_transformer, GaussianProcessRegressionDataTransformWrapper
+)
+class GPRwithDataTransforms(GaussianProcessRegression):
+    pass
+
+model = GPRwithDataTransforms(
+    model=build_gp_model(normalized_data),
+    optimizer=Optimizer(
+        gpflow.optimizers.Scipy(),
+        minimize_args={"options": dict(maxiter=100)}
+    ),
+    num_kernel_samples=100
+)
+
+
+# %% [markdown]
+#
+# Now we optimize and display the results.
+
+# %%
+# 30 * num_dims - num_initial_points
+num_steps = 56
+
+np.random.seed(1794)
+tf.random.set_seed(1794)
+
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+result = bo.optimize(num_steps, initial_data, model)
+dataset = result.try_get_final_dataset()
+
+query_points = dataset.query_points.numpy()
+observations = dataset.observations.numpy()
+
+arg_min_idx = tf.squeeze(tf.argmin(observations, axis=0))
+
+print(f"query point: {query_points[arg_min_idx, :]}")
+print(f"observation: {observations[arg_min_idx, :]}")
+
+def plot_regret_with_min(dataset):
+    observations = dataset.observations.numpy()
+    arg_min_idx = tf.squeeze(tf.argmin(observations, axis=0))
+
+    suboptimality = observations - F_MINIMUM.numpy()
+    ax = plt.gca()
+    plot_regret(suboptimality, ax, num_init=num_initial_points, idx_best=arg_min_idx)
+
+    ax.set_yscale("log")
+    ax.set_ylabel("Regret")
+    ax.set_ylim(0.001, 100000)
+    ax.set_xlabel("# evaluations")
+
+plot_regret_with_min(dataset)
+
+
+# %% [markdown]
+# ## Updating normalization and model parameters every iteration
+#
+# We now show how to modify the above to add methods to update model and normalization parameters. To achieve this, the user simply needs to define a method ``_update_model_and_normalization_parameters``.
+
+# %%
+class DynamicDataTransformModelWrapper(GaussianProcessRegressionDataTransformWrapper):
+    """DataTransformWrapper for a GaussianProcessRegression model."""
+
+    def _update_model_and_normalization_parameters(self, dataset):
+        """Update the model and normalization parameters based on the new dataset.
+        i.e. Denormalize using the old parameters, and renormalize using parameters set from
+        the new dataset.
+
+        :param dataset: New, unnormalized, dataset.
+        """
+        unnormalized_mean = self._observation_transformer.inverse_transform(self._model.mean_function.c)
+        self._observation_transformer.set_parameters(dataset.observations)
+        self._model.mean_function.c.assign(self._observation_transformer.transform(unnormalized_mean))
```

Comment:
Doesn't this assume a constant mean function?

Reply:
Yes, it does assume constant mean function. In general, updating the normalisation and model parameters like this will depend on the model, which is why this functionality needs to be specified by the user rather than being something we provide a standard implementation for.